src/simulations/collect.jl

```diff
@@ -164,6 +165,121 @@ function run!(
     return df_agent, df_model
 end
 
+"""
+    offline_run!(model, agent_step! [, model_step!], n::Integer; kwargs...) → agent_df, model_df
+    offline_run!(model, agent_step!, model_step!, n::Function; kwargs...) → agent_df, model_df
+
+Run the model while dumping data to CSV files. See [`run!`](@ref) for details.
+
+## Keywords
+* `writing_interval=1` : write to file every `writing_interval` times the data is collected. Small values have larger effects on performance.
+* `adata_savefile="adata.csv"` : a file to write agent data on.
+* `mdata_savefile="mdata.csv"`: a file to write the model data on.
+"""
+function offline_run! end
+
+offline_run!(model::ABM, agent_step!, n::Int = 1; kwargs...) =
+    offline_run!(model::ABM, agent_step!, dummystep, n; kwargs...)
+
+function offline_run!(
+    model,
+    agent_step!,
+    model_step!,
+    n;
+    when = true,
+    when_model = when,
+    mdata = nothing,
+    adata = nothing,
+    obtainer = identity,
+    agents_first = true,
+    showprogress = false,
+    writing_interval = 1,
+    adata_savefile = "adata.csv",
+    mdata_savefile = "mdata.csv"
+)
+
+    df_agent = init_agent_dataframe(model, adata)
+    df_model = init_model_dataframe(model, mdata)
+    if n isa Integer
+        if when == true
+            for c in eachcol(df_agent)
+                sizehint!(c, n)
+            end
+        end
+        if when_model == true
+            for c in eachcol(df_model)
+                sizehint!(c, n)
+            end
+        end
+    end
+
+    close(open(adata_savefile, "w"))
+    close(open(mdata_savefile, "w"))
+    model_appendtocsv = false
+    agent_appendtocsv = false
+
+    agent_count_collections = 0
+    model_count_collections = 0
+    agent_collected = false
+    model_collected = false
+
+    s = 0
+    p = if typeof(n) <: Int
+        ProgressMeter.Progress(n; enabled=showprogress, desc="run! progress: ")
+    else
+        ProgressMeter.ProgressUnknown(desc="run! steps done: ", enabled=showprogress)
+    end
+    while until(s, n, model)
+        if should_we_collect(s, model, when)
+            collect_agent_data!(df_agent, model, adata, s; obtainer)
+            agent_count_collections += 1
+            agent_collected = true
+        end
+        if should_we_collect(s, model, when_model)
+            collect_model_data!(df_model, model, mdata, s; obtainer)
+            model_count_collections += 1
+            model_collected = true
+        end
+        step!(model, agent_step!, model_step!, 1, agents_first)
+        s += 1
+
+        if model_collected && model_count_collections % writing_interval == 0
+            AgentsIO.CSV.write(mdata_savefile, df_model, append=model_appendtocsv)
+            df_model = init_model_dataframe(model, mdata)
```

Comment:
We re-initialize the data frames all the time that makes this rather expensive. What we should be doing is instead emptying the data frame and adding a new column each time. Emptying also retains column names which do not change during the run.

Reply:
Additionally, we should separate this using a function barrier. One function initializes the dataframes and the other adds 1 row and removes 1 row in place. So one function is just the loop itself.

Reply:
Not sure I get what you mean here. I now replaced the reinitialization of the dataframes with a call to `empty!`, and the data collection is just performed exactly in the same way as it is performed in the `run!` function.
There is no "adding one row and removing one in place"; data is collected as usual, but when the dumping is triggered we write the current df to file and empty it.